### PR TITLE
New blog post covering PostgreSQL unit tests for user-defined functions using pgTAP

### PR DESCRIPTION
**Pull request by TheoDimov**
Quick tutorial based on creating unit tests and using pgTAP to test user-defined stored functions and their results.

### CODE REVIEW

File: 2023/09/testing-user-defined-functions-with-pgtap.md [hunk 1]
```diff
@@ -0,0 +1,462 @@
...
+:schema
+    Name of a schema in which to find functions.
+:functions
+    An array of function names.
+:description
+    A short description of the test.
+```
+#### `ok()` and `is()`
+
+`ok()` and `is()` have similar a function with slightly different parameters:
```
Comment (vincepmartin) [inline 2023/09/testing-user-defined-functions-with-pgtap.md:108]: Maybe this should say 
`ok()` and `is()` have similar functionality with slightly different parameters

File: 2023/09/testing-user-defined-functions-with-pgtap.md [hunk 2]
```diff
@@ -0,0 +1,462 @@
...
+```postgresql
+SELECT is(add_numbers(5,5), 10, 'add_numbers correctly works');
+```
+
+You can find functions that cover pretty much anything in PostgreSQL here on [pgTAP's documentation site](https://pgtap.org/documentation.html) 
+
+
+### Testing more complex user-defined stored functions:
+
+An existing client product of ours, utilizes a specific function that also calls several other functions depending on its given argument. This function is called by the application
```
Comment (vincepmartin) [inline 2023/09/testing-user-defined-functions-with-pgtap.md:136]: Probably should remove the comma after "product of ours"

File: 2023/09/testing-user-defined-functions-with-pgtap.md (hunk 1 above)
Comment (vincepmartin) [inline 2023/09/testing-user-defined-functions-with-pgtap.md:108]: Maybe this should say
ok() and is() have similar functionality with slightly different parameters

File: 2023/09/testing-user-defined-functions-with-pgtap.md [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+and is passed an INTEGER as an argument from which the function bases an initial SELECT query’s WHERE condition on, in this case `WHERE id = [argument]’. Variables are then assigned
+column values from this queried table. These variables are then used as the conditions to `PERFORM` functions that change parts of the database.
+
+When testing this sort of scenario, it became apparent that I needed a set of test data, an ability to track that test data, run the function with the data, and test to see if the
```
Comment (vincepmartin) [inline 2023/09/testing-user-defined-functions-with-pgtap.md:140]: Possible reword?
```
When testing this sort of scenario, it became apparent that I needed a set of test data, the ability to track that test data, run the function with the data, and test to see if the changes are performed correctly or incorrectly.
```

No biggy.

File: 2023/09/testing-user-defined-functions-with-pgtap.md [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+column values from this queried table. These variables are then used as the conditions to `PERFORM` functions that change parts of the database.
+
+When testing this sort of scenario, it became apparent that I needed a set of test data, an ability to track that test data, run the function with the data, and test to see if the 
+changes are correctly, or incorrectly, performed. 
+
+The nice thing about pgTAP is that unit tests are meant to be rolled back leaving the database as you found it. When faced with a more complex function that nests other functions which
+loop through multiple tables, it may make more sense to fill those tables temporarily with the data required for the tests. 
+
+Let's take a nostalgia trip and imagine we’re operating a DVD rental store. The Pagila sample database, which you can download on GitHub, represents our scenario. It comes populated
+and we’ll construct a function which utilizes other functions which perform changes on multiple tables.
```
Comment (vincepmartin) [inline 2023/09/testing-user-defined-functions-with-pgtap.md:147]: "which utilizes other functions which perform changes on multiple tables" -> "which utilizes other functions that perform changes on multiple tables"

File: 2023/09/testing-user-defined-functions-with-pgtap.md [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+the references to `store` and `address`, so please take note in your own unit tests to be careful when referring to existing data.
+
+After we’ve established our procedure, we’re ready to run the tests. The pgTAP function `plan()` defines for pgTAP how many unit tests we are going to run. For the first example we run
+two tests. First we call the procedure `fill_test_data` which populates the database with test data and the temporary table to track that data. 
+
+Our first test is to check if the function exists. pgTAP’s `has_function` function is perfect for this. Pretty self explanatory as a function and more details can be found in the introduction
+of this article. 
+
+Our procedure is then called with a date for the test rental. This test rental’s id is saved into the temporary `test_instance_references` table which is then passed to the main function we
+want to test, `extension_check`. We call this function with the SELECT command, and immediately test the outcome with pgTAP’s `is()`. Our `:have` parameter holds to the return_date of our
```
Comment (vincepmartin) [inline 2023/09/testing-user-defined-functions-with-pgtap.md:317]: Maybe uneeded comma.

**Review by vincepmartin — COMMENTED** [review]
Good article!

**Comment by sethjensen1** [comment]
@TheoDimov This is live. Good post, thank you!

**Pull request closed without merging**